The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8272262376/job/22633763328) [(plain)](https://github.com/rust-lang/rust/commit/ed3e39ff4da8276c9f6dce161c328aea4aa5cb62/checks/22633763328/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_c8077930-f733-4ba9-b857-4b5b2bea32c0
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fix-step-forward-unchecked
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_c8077930-f733-4ba9-b857-4b5b2bea32c0
GITHUB_REF=refs/pull/122461/merge
GITHUB_REF_NAME=122461/merge
GITHUB_REF_PROTECTED=false
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#10 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#10 DONE 0.0s

#11 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#11 0.614   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#11 0.629 Collecting boolean-py==4.0
#11 0.638   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#11 0.655 Collecting chardet==5.1.0
---
#11 3.766 Building wheels for collected packages: reuse
#11 3.767   Building wheel for reuse (pyproject.toml): started
#11 4.097   Building wheel for reuse (pyproject.toml): finished with status 'done'
#11 4.098   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#11 4.098   Stored in directory: /tmp/pip-ephem-wheel-cache-gabk3z3p/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#11 4.101 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#11 4.123   Attempting uninstall: setuptools
#11 4.124     Found existing installation: setuptools 59.6.0
#11 4.125     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#11 5.369   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#11 5.544      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 21.8 MB/s eta 0:00:00
#11 5.596 Collecting filelock<4,>=3.12.2
#11 5.604   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 5.635 Collecting platformdirs<5,>=3.9.1
#11 5.642   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#11 5.664 Collecting distlib<1,>=0.3.7
#11 5.671   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#11 5.683      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 46.8 MB/s eta 0:00:00
#11 5.767 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#11 5.930 Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.2.0 virtualenv-20.25.1
#11 DONE 6.0s

#12 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#12 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     8189952 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/3cbb93223f33024db464a4df27a13c7cce870173/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-3cbb93223f33024db464a4df27a13c7cce870173-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 27.49s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/core/src/iter/range.rs at line 184:
 }
 
-
 // Separate impls for signed ranges because the distance within a signed range can be larger
 // Separate impls for signed ranges because the distance within a signed range can be larger
 // than the signed::MAX value. THerefore `as` casting to the signed type would be incorrect.
 macro_rules! step_signed_methods {
##[error]Diff in /checkout/library/core/src/iter/range.rs at line 200:
             // SAFETY: the caller has to guarantee that `start - n` doesn't overflow.
             unsafe { start.checked_sub_unsigned(n as $unsigned).unwrap_unchecked() }
-    }
+    };
 }
 
 
 macro_rules! step_unsigned_methods {
##[error]Diff in /checkout/library/core/src/iter/range.rs at line 216:
             // SAFETY: the caller has to guarantee that `start - n` doesn't overflow.
             unsafe { start.unchecked_sub(n as Self) }
-    }
+    };
 }
 
 
 // These are still macro-generated because the integer literals resolve to different types.
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/async_iter/async_iter.rs" "/checkout/library/core/src/async_iter/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/evaluate_obligation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/coherence.rs" "/checkout/library/core/src/hash/mod.rs" "/checkout/library/core/src/hash/sip.rs" "/checkout/library/core/src/ascii.rs" "/checkout/library/core/src/error.rs" "/checkout/library/core/src/any.rs" "/checkout/library/core/src/ascii/ascii_char.rs" "/checkout/library/core/src/cmp.rs" "/checkout/library/core/src/asserting.rs" "/checkout/library/core/src/clone.rs" "/checkout/library/core/src/time.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/infer_ctxt_ext.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/ambiguity.rs" "/checkout/compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs" "/checkout/compiler/rustc_trait_selection/src/traits/structural_match.rs" "/checkout/compiler/rustc_trait_selection/src/traits/object_safety.rs" "/checkout/compiler/rustc_trait_selection/src/traits/util.rs" "/checkout/compiler/rustc_trait_selection/src/traits/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/vtable.rs" "/checkout/compiler/rustc_trait_selection/src/traits/fulfill.rs" "/checkout/compiler/rustc_trait_selection/src/traits/misc.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/confirmation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs" "/checkout/compiler/rustc_trait_selection/src/traits/engine.rs" "/checkout/compiler/rustc_trait_selection/src/traits/outlives_bounds.rs" "/checkout/compiler/rustc_trait_selection/src/traits/project.rs" "/checkout/compiler/rustc_trait_selection/src/errors.rs" "/checkout/compiler/rustc_trait_selection/src/infer.rs" "/checkout/compiler/rustc_trait_selection/src/lib.rs" "/checkout/compiler/rustc_trait_selection/src/regions.rs" "/checkout/library/core/src/iter/range.rs" "/checkout/compiler/rustc_feature/src/removed.rs" "/checkout/compiler/rustc_feature/src/accepted.rs" "/checkout/compiler/rustc_feature/src/tests.rs" "/checkout/compiler/rustc_feature/src/unstable.rs" "/checkout/compiler/rustc_feature/src/builtin_attrs.rs" "/checkout/compiler/rustc_feature/src/lib.rs" "/checkout/library/core/src/iter/adapters/inspect.rs" "/checkout/library/core/src/iter/adapters/array_chunks.rs" "/checkout/library/core/src/iter/adapters/map_windows.rs" "/checkout/library/core/src/iter/adapters/scan.rs" "/checkout/library/core/src/iter/adapters/take.rs" "/checkout/library/core/src/iter/adapters/cycle.rs" "/checkout/library/core/src/iter/adapters/chain.rs" "/checkout/library/core/src/iter/adapters/flatten.rs" "/checkout/library/core/src/iter/adapters/step_by.rs" "/checkout/library/core/src/iter/adapters/by_ref_sized.rs" "/checkout/library/core/src/iter/adapters/mod.rs" "/checkout/library/core/src/iter/adapters/filter_map.rs" "/checkout/library/core/src/iter/adapters/filter.rs" "/checkout/library/core/src/iter/adapters/take_while.rs" "/checkout/library/core/src/iter/adapters/intersperse.rs" "/checkout/library/core/src/iter/adapters/skip.rs" "/checkout/library/core/src/iter/adapters/cloned.rs" "/checkout/library/core/src/iter/adapters/enumerate.rs" "/checkout/library/core/src/iter/adapters/map.rs" "/checkout/library/core/src/async_iter/from_iter.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Mar 13 22:18:06 UTC 2024
  network time: Wed, 13 Mar 2024 22:18:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/core/src/iter/range.rs
+++ b/library/core/src/iter/range.rs
@@ -184,8 +184,25 @@ pub trait Step: Clone + PartialOrd + Sized {
     }
 }
 
-// These are still macro-generated because the integer literals resolve to different types.
-macro_rules! step_identical_methods {
+// Separate impls for signed ranges because the distance within a signed range can be larger
+// than the signed::MAX value. Therefore `as` casting to the signed type would be incorrect.
+macro_rules! step_signed_methods {
+    ($unsigned: ty) => {
+        #[inline]
+        unsafe fn forward_unchecked(start: Self, n: usize) -> Self {
+            // SAFETY: the caller has to guarantee that `start + n` doesn't overflow.
+            unsafe { start.checked_add_unsigned(n as $unsigned).unwrap_unchecked() }
+        }
+
+        #[inline]
+        unsafe fn backward_unchecked(start: Self, n: usize) -> Self {
+            // SAFETY: the caller has to guarantee that `start - n` doesn't overflow.
+            unsafe { start.checked_sub_unsigned(n as $unsigned).unwrap_unchecked() }
+        }
+    };
+}
+
+macro_rules! step_unsigned_methods {
     () => {
         #[inline]
         unsafe fn forward_unchecked(start: Self, n: usize) -> Self {
@@ -198,7 +215,12 @@ macro_rules! step_identical_methods {
             // SAFETY: the caller has to guarantee that `start - n` doesn't overflow.
             unsafe { start.unchecked_sub(n as Self) }
         }
+    };
+}
 
+// These are still macro-generated because the integer literals resolve to different types.
+macro_rules! step_identical_methods {
+    () => {
         #[inline]
         #[allow(arithmetic_overflow)]
         #[rustc_inherit_overflow_checks]
@@ -239,6 +261,7 @@ macro_rules! step_integer_impls {
             #[unstable(feature = "step_trait", reason = "recently redesigned", issue = "42168")]
             impl Step for $u_narrower {
                 step_identical_methods!();
+                step_unsigned_methods!();
 
                 #[inline]
                 fn steps_between(start: &Self, end: &Self) -> Option<usize> {
@@ -271,6 +294,7 @@ macro_rules! step_integer_impls {
             #[unstable(feature = "step_trait", reason = "recently redesigned", issue = "42168")]
             impl Step for $i_narrower {
                 step_identical_methods!();
+                step_signed_methods!($u_narrower);
 
                 #[inline]
                 fn steps_between(start: &Self, end: &Self) -> Option<usize> {
@@ -335,6 +359,7 @@ macro_rules! step_integer_impls {
             #[unstable(feature = "step_trait", reason = "recently redesigned", issue = "42168")]
             impl Step for $u_wider {
                 step_identical_methods!();
+                step_unsigned_methods!();
 
                 #[inline]
                 fn steps_between(start: &Self, end: &Self) -> Option<usize> {
@@ -360,6 +385,7 @@ macro_rules! step_integer_impls {
             #[unstable(feature = "step_trait", reason = "recently redesigned", issue = "42168")]
             impl Step for $i_wider {
                 step_identical_methods!();
+                step_signed_methods!($u_wider);
 
                 #[inline]
                 fn steps_between(start: &Self, end: &Self) -> Option<usize> {

--- a/library/core/tests/iter/range.rs
+++ b/library/core/tests/iter/range.rs
@@ -325,6 +325,11 @@ fn test_range_advance_by() {
     assert_eq!(Ok(()), r.advance_back_by(usize::MAX));
 
     assert_eq!((r.start, r.end), (0u128 + usize::MAX as u128, u128::MAX - usize::MAX as u128));
+
+    // issue 122420, Step::forward_unchecked was unsound for signed integers
+    let mut r = -128i8..127;
+    assert_eq!(Ok(()), r.advance_by(200));
+    assert_eq!(r.next(), Some(72));
 }
 
 #[test]

--- a/src/tools/miri/tests/pass/shims/time-with-isolation2.stdout
+++ b/src/tools/miri/tests/pass/shims/time-with-isolation2.stdout
@@ -1,1 +1,1 @@
-The loop took around 7s
+The loop took around 12s

